cmds/bootcm4.c

```diff
@@ -0,0 +1,179 @@
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <devices/devs.h>
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <phfs/phfs.h>
+
+
+#define CM4_BOOT_MEMSIZE 0x00040000u
+#define CM4_BOOT_ADDRESS 0x20200000u
+#define CM4_VTOR_ADDRESS 0x20200100u
+
+
+static unsigned int isImgLoaded = 0u;
+
+
+static void cmd_bootcm4Info(void)
+{
+	lib_printf("boots CM4 core image, use <-h> to see usage");
+}
+
+
+static void print_usage(const char *name)
+{
+	lib_printf(
+		"Usage: %s <options> [<device image>]\n"
+		"\t-b  boot CM4 core\n"
+		"\t-l  load binary image using phfs <device> and <image>",
+		name);
+}
+
+
+static int loadImagePhfs(const char *phfsDev, const char *phfsFile)
+{
+	u8 buff[SIZE_MSG_BUFF];
+	handler_t h;
+	addr_t offs = 0u;
+	addr_t addr = CM4_BOOT_ADDRESS;
+
+	int res = phfs_open(phfsDev, phfsFile, 0, &h);
+
+	if (res < 0) {
+		log_error("Can't open '%s' on '%s'", phfsFile, phfsDev);
+		return res;
+	}
+
+	do {
+		res = phfs_read(h, offs, buff, sizeof(buff));
+		if (res < 0) {
+			log_error("Can't read '%s' from '%s'", phfsFile, phfsDev);
+			break;
+		}
+
+		hal_memcpy((void *)addr, buff, res);
+		offs += res;
+		addr += res;
+
+		if (offs > CM4_BOOT_MEMSIZE) {
+			log_error("Image '%s' too big", phfsFile);
+			res = -ENOSPC;
+			break;
+		}
+	} while (res > 0);
+
+	phfs_close(h);
+
+	return res;
+}
+
+static int cmd_bootcm4(int argc, char *argv[])
+{
+	const char *phfsDev = NULL;
+	const char *phfsFile = NULL;
+
+	int opt;
+	int optLoad = 0;
+	int optBoot = 0;
+
+	lib_printf("\n");
+
+	if (argc == 1) {
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	for (;;) {
+		opt = lib_getopt(argc, argv, "bl");
+		if (opt == -1) {
+			break;
+		}
+
+		switch (opt) {
+			case 'b':
+				optBoot = 1;
+				break;
+
+			case 'l':
+				optLoad = 1;
+				break;
+
+			case 'h':
+			default:
+				print_usage(argv[0]);
+				return CMD_EXIT_FAILURE;
+		}
+	}
+
+	if (argc > optind) {
+		if (((argc - optind) != 2) || (optLoad == 0)) {
+			log_error("%s: Invalid arguments", argv[0]);
+			return CMD_EXIT_FAILURE;
+		}
+		phfsDev = argv[optind++];
+		phfsFile = argv[optind++];
+	}
+
+	if (_imxrt_getStateCM4() != 0u) {
+		if ((optLoad != 0) || (optBoot != 0)) {
+			log_error("%s: CM4 core is already running", argv[0]);
+			return CMD_EXIT_FAILURE;
+		}
+	}
+
+	if (optLoad != 0) {
+		if ((phfsDev == NULL) || (phfsFile == NULL)) {
+			log_error("%s: Missing argument <device file> to load", argv[0]);
+			return CMD_EXIT_FAILURE;
+		}
+
+		if (loadImagePhfs(phfsDev, phfsFile) < 0) {
+			return CMD_EXIT_FAILURE;
+		}
+
+		isImgLoaded++;
+	}
+
+	if (optBoot != 0) {
+		if (isImgLoaded == 0u) {
+			log_error("%s: CM4 core image not loaded", argv[0]);
+			return CMD_EXIT_FAILURE;
+		}
+
+		if (_imxrt_setVtorCM4(0, 0, CM4_VTOR_ADDRESS) < 0) {
+			log_error("%s: Unable to init CM4 core vectors", argv[0]);
+			return CMD_EXIT_FAILURE;
+		}
+
+		hal_invalDCacheAddr((void *)CM4_BOOT_ADDRESS, CM4_BOOT_MEMSIZE);
```

Comment:
Seems ok, done.

Reply:
No, at the end of image loading block.

Reply:
Sorry for misunderstanding. Cache needs to be coherent right before CM4 reset release (src_scr|=1, see reference manual) and (seems ok to me) before vectors are set, that's it. There is no need to clear cache after loading CM4 image in plo. If you meant (I guess) to use plo only to load the CM4 image and boot already loaded image using slightly modified `cm4-tool` from the user-space it will do just fine, because whole cache is invalidated in [`hal_done()`](https://github.com/phoenix-rtos/plo/blob/fb45e0f00e6818a884d6440790d7b1dfd034df2d/hal/armv7m/imxrt/hal.c#L63) just before [jump to](https://github.com/phoenix-rtos/plo/blob/fb45e0f00e6818a884d6440790d7b1dfd034df2d/cmds/go.c#L41-L42) the kernel. So there is no need to move this.

Reply:
Ok, done.